freqtrade/exchange/__init__.py

```diff
@@ -155,6 +157,12 @@ def id(self) -> str:
         """exchange ccxt id"""
         return self._api.id
 
+    def klines(self, pair: str) -> DataFrame:
+        if pair in self._klines:
+            return self._klines[pair].copy()
```

Comment:
just thinking loud ... every time analyze pair requests DF, we are returning a new copy of it. performance wise is it ok? but I guess we don't have any choice other than passing by reference. also each time analyze is called it is going through all indicators for all candles. even if process_only_new_candles is true, it analyzes all candles as soon as it sees one. I am thinking about if there is an option in pandas to avoid that. cause once an indicator is populated for a candle it would be set in stone. passing by reference would resolve that problem if pandas have this option but then it would become messy ... 
Anyway sorry it is not related to this PR.

Reply:
better/faster than creating a new dataframe each iteration, yes.

The danger with passing by reference is if you use the data somewhere else (say edge) ... you'll want to rely on the data not beeing tampered with from the exchange. i could shift all closes (by accident or on purpose) - and render the bot useless without ever really noticing and making it nearly impossible to detect.

Also pandas has a very odd way to handle references / assignments - where the autors themselfs state they cannot guarantee when a copy or a view of the dataframe is used (i guess it depends on the content / size?) - at least that's what i read somewhere.

We need to process all candles again when a new candle arrives - even something as simple as SMA would require all candles (within the set period) to calculate the results - and by using ta-lib it's quite fast compared to other solutions.
I tried `pyti` recently and didn't have the patience to wait until bollingers calculation completed for 2 months and 6 pairs before interrupting ...

Additionally caching populated candles (with indicators) could be an option - but i didn't think about that yet.

Reply:
fair enough !